### ISSUE TITLE
[Feature] Receipt trait in alloy-network

### ISSUE THREAD

**Issue by prestwich** [alloy-rs/alloy#466]
Followup work: 
- `trait Receipt` in alloy-network
- Receipt bound on `type Recept` in `trait Network`
- `fn status` in `trait Receipt`
- `impl Receipt for ReceiptEnvelope` and related types
- change this function to a delegate to `Receipt`

_Originally posted by @prestwich in https://github.com/alloy-rs/alloy/pull/456#discussion_r1552786585_

**Comment by developeruche** [comment]
@prestwich can you give more context to;

Receipt bound on `type Recept` in `trait Network`

and 

`impl<T: Receipt> Receipt for ReceiptResponse`

**Comment by zilayo** [comment]
seems related so leaving here instead of a new issue.

Current UX for accessing logs from a `WithOtherFields<TransactionReceipt<AnyReceiptEnvelope<Log>>>` receipt is:

`receipt.inner.inner.inner.receipt.logs`

Currently there are only implementations for `AnyReceiptEnvelope<()>` - these need to be exposed for `AnyReceiptEnvelope<T>` as well. 

https://github.com/alloy-rs/alloy/blob/cd5e0f7397ed599ae384abbc282b73fbf6e8c16c/crates/consensus/src/receipt/any.rs#L18-L27

Note - there's already a `TxReceipt` trait that has `fn success` which is equivalent to what `fn status` would be.

Perhaps makes sense to remove this trait altogether in favor of the `Receipt` trait mentioned above?

**Comment by prestwich** [comment]
> Note - there's already a TxReceipt trait that has fn success which is equivalent to what fn status would be.

Yeah, i think my issue here isn't "create a Receipt trait" it's "ensure that the existing Receipt trait is properly implemented for relevant types further up the abstraction stack, like `ReceiptResponse` and `AnyReceiptEnvelope`, and used in abstracted code like the `Provider` and `PendingTransaction`

**Comment by prestwich** [comment]
> Current UX for accessing logs from a `WithOtherFields<TransactionReceipt<AnyReceiptEnvelope<Log>>>` receipt is:
> 
> `receipt.inner.inner.inner.receipt.logs`
> 
> Currently there are only implementations for `AnyReceiptEnvelope<()>` - these need to be exposed for `AnyReceiptEnvelope<T>` as well.

This is the sort of UX problem that a consistenyl applied and used `Receipt` trait would solve 👍

**Comment by developeruche** [comment]
> > Note - there's already a TxReceipt trait that has fn success which is equivalent to what fn status would be.
> 
> Yeah, i think my issue here isn't "create a Receipt trait" it's "ensure that the existing Receipt trait is properly implemented for relevant types further up the abstraction stack, like `ReceiptResponse` and `AnyReceiptEnvelope`, and used in abstracted code like the `Provider` and `PendingTransaction`

Oh yeah... got it
